[WIP] Optionally require logins for certificate views

kaizoku opened the pull request:
We'd like to optionally require logins for viewing and downloading certificates. This adds a waffle switch which when enabled uses the django `login_required`decorator on the certificate views.

This also adds the certificate status to the GeneratedCertificates admin page view.

**JIRA tickets**: Implements BB-1388.

**Dependencies**: None

**Sandbox URL**: TBD - sandbox is being provisioned.

**Testing instructions**:

1. Generate a certificate to test with.
    ```
    /edx/bin/{python,manage}.edxapp lms --settings=devstack_docker generate_fake_cert verified course-v1:edX+DemoX+Demo_Course
    ```
2. Check the certificate renders normally with no login required in an logged-out browser: http://localhost:18000/certificates/user/8/course/course-v1:edX+DemoX+Demo_Course
3. Enable the waffle switch
    ```
    /edx/bin/{python,manage}.edxapp lms --settings=devstack_docker waffle_switch certificates.require_login on
    ```
4. Check the certificate page again and ensure it requires login now.

**Author notes and concerns**:

1. Should we create a waffle configuration in a new file for the certificates application?

**Reviewers**
- [ ] (@giovannicimolin )
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @kaizoku! I've created [OSPR-3752](https://openedx.atlassian.net/browse/OSPR-3752) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@kaizoku Thank you for your contribution. Please let me know once tests are green and this is ready for our review.

kaizoku commented:
@natabene , I think that test failure was a fluke, it's an HTTP Connection error from selenium, and doesn't come from any components changed in this PR. The test works successfully on this branch in my devstack. I'll see if I can get jenkins to re-run it.

kaizoku commented:
jenkins run bokchoy

kaizoku commented:
Thanks @giovannicimolin!

Hm, perhaps we should use a `CourseWaffleFlag`, that would provide more fine grained control.

My worry there is that course authors would have to create a new waffle flag for every course they create, and authentication wouldn't be enforced if they hadn't gotten the memo about creating waffle flags for their course. Whereas an organization who wants authentication for their certificates would likely want them for all courses, so missing this on one course would maybe be a kind of breach.

Also for the documentation note, I'm not sure where waffle flags are typically documented.
I did see a [post today](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/edx-code/8wbJleGikPI/icFV9uwIBgAJ) about documentation of waffle flags, which references [OEP-17](https://open-edx-proposals.readthedocs.io/en/latest/oep-0017-bp-feature-toggles.html), but that OEP mostly describes when and why to create a waffle flag, but not much of how or where to document it.

kaizoku commented:
@giovannicimolin, I've switched this around to using a `CourseWaffleFlag`, since it defers to `WaffleFlag`, so we can still set a global, but have overrides for each course as needed.

giovannicimolin commented:
@kaizoku Nice work here! I tested it again and it works perfectly. 

:+1:

- [x]  I tested this:
1. Ran the same tests as before, but switching to a flag instead of a switch.
2. Checked that viewing certificates can require login either on a per course basis or instance wide using the `certificates.require_login`. :heavy_check_mark: 
- [x] I read through the code
- [ ] Includes documentation :warning:
- [ ] ~~I checked for accessibility issues~~ NA

giovannicimolin commented:
@natabene This is ready for edX review.

natabene commented:
@ormsbee This is ready for your review, when you have a chance.

natabene commented:
@marcotuts Do you want to review this before it goes to engineering?

kaizoku commented:
@marcotuts, @ormsbee any chance you've been able to check this out?

natabene commented:
@kaizoku Sorry for delay, we will try to look at this this week.

kaizoku commented:
No problem @natabene , thanks for the response :)

kaizoku commented:
@natabene, any chance you've been able to look at this PR?

natabene commented:
@kaizoku Sorry, not yet. We will try today.

kaizoku commented:
@natabene has this come up for review yet?

natabene commented:
@kaizoku Sorry, no. We now prioritize with @pomegranited , please reach out to her for coordination.

marcotuts commented:
On the product side I'm not sure I understand the value / goal of this contribution, but the change seems small enough here that I'm not necessarily keen on blocking this. 

I plan to move this to engineering review but this is a low priority review for us currently, to provide some context. 

For my purposes, could you help me understand what the primary driving goal of this feature is, if learners generally are in the driver's seat of whether or not they share their public certificates in the first place.

kaizoku commented:
@marcotuts since the certificate URLs are hard to guess, it is already mostly up to the learner if they share that URL.

This PR came out of our work trying to reduce code drift from a customized codebase we inherited, so I'm not entirely sure what the initial rational for this was except to provide another level of privacy to that. I can ask what the initial drive for this was though and see if there was something else behind it.

kaizoku commented:
@marcotuts, the codebase we're trying to reduce code drift from uses a custom plugin to provide certificates externally, so this is was meant so learners can view their own certificate page, but use the externally exported certificates for public attestation of their course completion.

giovannicimolin commented:
jenkins run all

giovannicimolin commented:
jenkins run all

natabene commented:
jenkins run all

bradenmacdonald commented:
Note to self: when I get to review this, perhaps it should use the pluggable override proposed in #21433 to be even more generic; that's also a good test case for pluggable overrides.

marcotuts commented:
Moving this to product review - I'm still not clear on the value beyond code drift that this has for us to review/merge currently.

bradenmacdonald commented:
@marcotuts We are working to change the implementation here to a plugin approach so that there is no increase in code complexity to the core platform.

marcotuts commented:
Sounds good Braden I'll push this past product review back to you / cc-review. This is the kind of PR that adds complexity and whose value is unclear/specific to another instance that would have led to me recommending we close this. Glad to hear the plugin approach may still lead to progress here though!

kaizoku commented:
@pomegranited, we'll still need to add one line to enable the plugin point after https://github.com/edx/edx-platform/pull/21433 is merged.
This PR currently has most of the contents from that PR for testing.
Once it's merged though, I'll cut it down to the decorator which creates the pluggability point: `@pluggable_override('RENDER_CERTIFICATE_VIEW_IMPL')`

pomegranited commented:
No worries, thanks for the update @kaizoku !

Agrendalath commented on lms/djangoapps/certificates/views/webview.py:
```diff
@@ -473,10 +474,12 @@ def render_cert_by_uuid(request, certificate_uuid):
     """
     This public view generates an HTML representation of the specified user and course
     If a certificate is not available, we display a "Sorry!" screen instead
+    It can be overridden by setting `RENDER_CERTIFICATE_VIEW_IMPL` to an alternative implementation.
```
I see that the docstring, from which this has been copied, was not updated after the reworks. I'll fix this in #27466. Please use the prefix `OVERRIDE_` for this (e.g. [OVERRIDE_GET_UNIT_ICON](https://github.com/edx/edx-platform/blob/068ac5290f0e5dcd770b789d766b2cd5d22a1376/openedx/core/lib/xblock_utils/__init__.py#L558)) to keep it consistent.

kaizoku replied:
Thanks for pointing that out. I've updated the variable name here to match.

kaizoku commented:
This PR is ready for upstream review.

bradenmacdonald reviewed (approved):
@kaizoku Code looks good to me 👍🏻 

Can you please: (1) Update the PR description to match the final approach taken here, and (2) rebase+squash to one commit with a [conventional commit](https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0051-bp-conventional-commits.rst) prefix?

kaizoku commented:
Thanks for the review @bradenmacdonald. I've squashed the PR, updated the commit message, and updated the PR description here. This should be ready now.

bradenmacdonald commented:
jenkins run js

edx-webhook commented:
@kaizoku 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21147&entry.1484655318=https://github.com/kaizoku&entry.752974735=2019-07-17+08:32&entry.1917517419=2021-05-28+18:10&entry.2133058324=Yes) so we can improve your experience in the future.

kaizoku commented:
Thanks for the review @bradenmacdonald !
